File: lib/lint.ts
import fs from "fs";
import path from "path";
import matter from "gray-matter";
import { CATEGORY_DIRS, HUB, MOVIES_DIR, TASTE_DIR, mdFiles, wikilinks } from "./vault";

const REPORT = path.join(process.cwd(), "data/lint-pending.txt");

// mechanical, deterministic fixes applied in place; returns what was fixed
// ponytail: hub-only, wikilink syntax only — widen scope when the report shows
// recurring patterns elsewhere
export function autofixVault(): string[] {
  if (!fs.existsSync(HUB)) return [];
  const before = fs.readFileSync(HUB, "utf8");
  const after = before
    // markdown formatting inside a wikilink target: [[**Foo**]] → [[Foo]]
    .replace(/\[\[\s*[*_`]+([^\]|#]*?)[*_`]+\s*\]\]/g, (_, t) => `[[${t.trim()}]]`)
    // unclosed wikilink: [[Foo] → [[Foo]]
    .replace(/\[\[([^\][\n]+)\](?!\])/g, "[[$1]]");
  if (after === before) return [];
  fs.writeFileSync(HUB, after);
  return ["hub: normalized malformed wikilink syntax"];
}

// judgment calls for the agent: dead links, orphan pages, unlinked taste bullets
export function lintVault(): string[] {
  if (!fs.existsSync(HUB)) return [];
  const issues: string[] = [];
  const pageFiles = [...mdFiles(MOVIES_DIR), ...CATEGORY_DIRS.flatMap(mdFiles)];
  const pages = new Set(pageFiles.map((f) => path.basename(f, ".md")));
  // taste/ pages (deep profile) resolve links but aren't category/movie pages
  for (const f of mdFiles(TASTE_DIR)) pages.add(path.basename(f, ".md"));
  const hub = matter(fs.readFileSync(HUB, "utf8")).content;

  // watchlist and not-interested bullet titles legitimately have no page yet —
  // exclude from the dead-link check
  const pageless = new Set<string>();
  const sections = hub.split(/^## /m);
  for (const section of sections) {
    const heading = section.split("\n", 1)[0].toLowerCase();
    const bullets = section
      .split("\n")
      .filter((l) => /^\s*-\s+\S/.test(l) && !l.includes("(empty"))
      .map((l) => l.replace(/^\s*-\s+/, "").trim());
    if (heading.startsWith("watchlist") || heading.startsWith("not interested")) {
      for (const b of bullets) {
        const first = wikilinks(b)[0];
        if (first) pageless.add(first);
      }
    } else if (heading.startsWith("taste")) {
      for (const b of bullets) {
        if (wikilinks(b).length === 0)
          issues.push(`taste bullet has no [[Category]] links: "${b.slice(0, 80)}"`);
      }
    }
  }

  const linked = new Set<string>();
  const scan = (text: string, where: string) => {
    for (const target of wikilinks(text)) {
      linked.add(target);
      if (!pages.has(target) && !pageless.has(target))
        issues.push(`dead wikilink [[${target}]] in ${where}`);
    }
  };
  scan(hub, "hub");
  for (const f of pageFiles) scan(fs.readFileSync(f, "utf8"), path.basename(f));

  for (const p of pages) {
    if (!linked.has(p)) issues.push(`orphan page (no inbound links): ${p}.md`);
  }

  issues.push(...lintIndexes());

  // dedupe (same dead link can appear in many files) and cap the report
  return [...new Set(issues)].slice(0, 20);
}

// hierarchical-index invariants: each dimension _index.md mirrors its directory,
// a page is indexed in exactly one dimension, sub-indexes stay a sane size, and
// the grand index routes to every dimension (container rule)
function lintIndexes(): string[] {
  const issues: string[] = [];
  const grandFile = path.join(MOVIES_DIR, "_index.md");
  const grand = fs.existsSync(grandFile) ? fs.readFileSync(grandFile, "utf8") : "";
  const seen = new Map<string, string>(); // page -> dimension it's indexed under

  for (const dir of CATEGORY_DIRS) {
    const dim = path.basename(dir);
    const indexFile = path.join(dir, "_index.md");
    if (!fs.existsSync(dir)) continue;
    if (!fs.existsSync(indexFile)) {
      issues.push(`missing sub-index: wiki/movies/${dim}/_index.md`);
      continue;
    }
    const onDisk = new Set(mdFiles(dir).map((f) => path.basename(f, ".md")));
    // one entry per "- [[Page]] — ..." bullet; first wikilink names the page
    const indexed = new Set(
      fs
        .readFileSync(indexFile, "utf8")
        .split("\n")
        .filter((l) => /^-\s+\[\[/.test(l))
        .map((l) => wikilinks(l)[0])
        .filter(Boolean)
    );
    for (const p of onDisk)
      if (!indexed.has(p)) issues.push(`${dim}/_index.md is missing its page [[${p}]]`);
    for (const p of indexed) {
      if (!onDisk.has(p)) issues.push(`${dim}/_index.md lists [[${p}]] but ${dim}/${p}.md does not exist`);
      else if (seen.has(p)) issues.push(`[[${p}]] indexed in both ${seen.get(p)}/ and ${dim}/ (must be exactly one)`);
      else seen.set(p, dim);
    }
    // hard ceiling only — the ~40-entry split judgment belongs to the Flow 3
    // consolidation pass, not a per-turn nag
    if (indexed.size > 50)
      issues.push(`${dim}/_index.md has ${indexed.size} entries (>50) — needs a dimension split at next consolidation`);
    if (grand && !grand.includes(`movies/${dim}/_index`))
      issues.push(`grand index wiki/movies/_index.md has no row for dimension ${dim}/`);
  }
  return issues;
}

export function writeLintReport(issues: string[]) {
  fs.mkdirSync(path.dirname(REPORT), { recursive: true });
  if (issues.length) fs.writeFileSync(REPORT, issues.join("\n"));
  else fs.rmSync(REPORT, { force: true });
}

export function consumeLintReport(): string {
  if (!fs.existsSync(REPORT)) return "";
  const text = fs.readFileSync(REPORT, "utf8").trim();
  fs.rmSync(REPORT, { force: true });
  return text;
}

File: jetson/payload/gardener/lint.py
"""Vault linter. Universal rules run on any vault; hub rules are
profile-gated. The movie profile reproduces lib/lint.ts rule-for-rule
(hub-only autofix scope, watchlist-pageless exclusion, taste-bullet links,
dedupe, cap of 20).

Universal (always):
  dead_link        [[Target]] with no page (minus hub pageless sections);
                   path-style targets like [[movies/_index]] count as live
                   when the file exists in the vault — Obsidian resolves
                   them, and a false dead-link here would become a queued
                   "repair" that damages a healthy hub
  orphan           page with no inbound links
  duplicate_name   two pages share a basename; Obsidian links resolve to one

Hub-gated (profile.hub set):
  hub dead links, hub-inbound credit for orphans, pageless sections
  (PAGELESS_SECTIONS), and <section>_unlinked for LINKED_BULLET_SECTIONS
  bullets carrying no [[link]] (movie profile: taste_unlinked).

Index-invariant rules (profile.category_dirs set — port of lib/lint.ts
lintIndexes): each dimension _index.md mirrors its directory exactly, a page
is indexed in exactly one dimension, sub-indexes stay under 50 entries, and
the grand index (GRAND_INDEX) routes to every dimension. Kind: index_mismatch.

Only dead_link and orphan ever become queue items; the report-only kinds
surface in `gardener lint` and the Sonnet audit evidence.
"""
import os
import re

from . import frontmatter
from .wikilink import wikilinks

LINT_CAP = 20

# [[**Foo**]] -> [[Foo]]  (markdown formatting inside a wikilink target)
_FORMATTED = re.compile(r"\[\[\s*[*_`]+([^\]|#]*?)[*_`]+\s*\]\]")
# [[Foo] -> [[Foo]]  (unclosed wikilink)
_UNCLOSED = re.compile(r"\[\[([^\][\n]+)\](?!\])")


def autofix_text(text):
    text = _FORMATTED.sub(lambda m: "[[%s]]" % m.group(1).strip(), text)
    text = _UNCLOSED.sub(r"[[\1]]", text)
    return text


def autofix_vault(vault):
    """Mechanical fixes, hub-only (parity with lib/lint.ts; a hub-less vault
    has no autofix in v1). Returns list of fix descriptions."""
    if not vault.hub or not os.path.exists(vault.hub):
        return []
    before = vault.read(vault.hub)
    after = autofix_text(before)
    if after == before:
        return []
    with open(vault.hub, "w", encoding="utf-8") as fh:
        fh.write(after)
    return ["hub: normalized malformed wikilink syntax"]


def _hub_sections(hub_body):
    """[(lowercased heading, [bullet texts])] for each '## ' section."""
    out = []
    for section in hub_body.split("\n## "):
        heading = section.split("\n", 1)[0].lower().lstrip("# ")
        bullets = [
            re.sub(r"^\s*-\s+", "", line).strip()
            for line in section.split("\n")
            if re.match(r"^\s*-\s+\S", line) and "(empty" not in line
        ]
        out.append((heading, bullets))
    return out


def lint_vault(vault):
    """Issues as dicts {kind, detail, target, where}; detail matches the
    upstream lint.ts message format for the shared kinds."""
    issues = []
    seen = set()

    def add(kind, detail, target="", where=""):
        if detail in seen:
            return
        seen.add(detail)
        issues.append(
            {"kind": kind, "detail": detail, "target": target, "where": where}
        )

    pages = vault.pages()

    for name, winner, loser in vault.duplicates:
        add(
            "duplicate_name",
            "duplicate page name '%s': %s shadows %s" % (name, winner, loser),
            target=name,
            where=loser,
        )

    # -- hub rules (profile-gated) --------------------------------------------
    pageless = set()
    hub_body = None
    if vault.hub and os.path.exists(vault.hub):
        hub_body = frontmatter.body(vault.read(vault.hub))
        profile = vault.profile
        for heading, bullets in _hub_sections(hub_body):
            if any(heading.startswith(s) for s in profile.pageless_sections):
                for b in bullets:
                    links = wikilinks(b)
                    if links:
                        pageless.add(links[0])
            elif any(heading.startswith(s) for s in profile.linked_bullet_sections):
                section = next(
                    s for s in profile.linked_bullet_sections if heading.startswith(s)
                )
                for b in bullets:
                    if not wikilinks(b):
                        add(
                            "%s_unlinked" % section,
                            '%s bullet has no [[Category]] links: "%s"'
                            % (section, b[:80]),
                            target=b[:80],
                            where="hub",
                        )

    # -- universal rules --------------------------------------------------------
    linked = set()
    # the hub is a real page even though it's never in pages() — [[Movies]]
    # from a taste page must not read as dead
    live = set(pages)
    if vault.hub:
        live.add(os.path.splitext(os.path.basename(vault.hub))[0])

    def resolves_as_path(target):
        """[[movies/_index]]-style links: Obsidian resolves them by path
        suffix anywhere in the vault, so they're live if any .md file's
        vault-relative path ends with the target."""
        if "/" not in target:
            return False
        rel = target if target.endswith(".md") else target + ".md"
        for dirpath, dirnames, filenames in os.walk(vault.root):
            dirnames[:] = [d for d in dirnames if not d.startswith(".")]
            for f in filenames:
                if not f.endswith(".md"):
                    continue
                full_rel = vault.relpath(os.path.join(dirpath, f))
                if full_rel == rel or full_rel.endswith(os.sep + rel):
                    return True
        return False

    def scan(text, where):
        for target in wikilinks(text):
            linked.add(target)
            if target in live or target in pageless:
                continue
            if resolves_as_path(target):
                continue
            add(
                "dead_link",
                "dead wikilink [[%s]] in %s" % (target, where),
                target=target,
                where=where,
            )

    if hub_body is not None:
        scan(hub_body, "hub")
    for path in sorted(pages.values(), key=lambda p: vault.relpath(p)):
        scan(vault.read(path), os.path.basename(path))

    for name in pages:
        if name not in linked:
            add(
                "orphan",
                "orphan page (no inbound links): %s.md" % name,
                target=name,
                where=vault.relpath(pages[name]),
            )

    for detail in _lint_indexes(vault):
        add("index_mismatch", detail)

    return issues[:LINT_CAP]


_INDEX_BULLET = re.compile(r"^-\s+\[\[")


def _index_entries(text):
    """Pages named by '- [[Page]] — ...' bullets (first wikilink per bullet)."""
    out = set()
    for line in text.split("\n"):
        if _INDEX_BULLET.match(line):
            links = wikilinks(line)
            if links:
                out.add(links[0])
    return out


def _lint_indexes(vault):
    """Port of lib/lint.ts lintIndexes: each dimension _index.md mirrors its
    directory, one dimension per page, size ceiling, grand-index routing."""
    profile = vault.profile
    if not profile.category_dirs:
        return []
    issues = []
    grand = ""
    if profile.grand_index:
        grand_path = vault.resolve(profile.grand_index)
        if grand_path and os.path.isfile(grand_path):
            grand = vault.read(grand_path)
    seen = {}  # page -> dimension it's indexed under
    for rel in profile.category_dirs:
        directory = os.path.join(vault.root, rel)
        dim = os.path.basename(rel.rstrip("/"))
        if not os.path.isdir(directory):
            continue
        index_file = os.path.join(directory, "_index.md")
        if not os.path.isfile(index_file):
            issues.append("missing sub-index: %s/_index.md" % rel)
            continue
        on_disk = {
            os.path.splitext(os.path.basename(f))[0]
            for f in vault.md_files(directory)
        }
        indexed = _index_entries(vault.read(index_file))
        for p in sorted(on_disk - indexed):
            issues.append("%s/_index.md is missing its page [[%s]]" % (dim, p))
        for p in sorted(indexed):
            if p not in on_disk:
                issues.append(
                    "%s/_index.md lists [[%s]] but %s/%s.md does not exist"
                    % (dim, p, dim, p)
                )
            elif p in seen:
                issues.append(
                    "[[%s]] indexed in both %s/ and %s/ (must be exactly one)"
                    % (p, seen[p], dim)
                )
            else:
                seen[p] = dim
        if len(indexed) > 50:
            issues.append(
                "%s/_index.md has %d entries (>50) — needs a dimension split "
                "at next consolidation" % (dim, len(indexed))
            )
        if grand and ("%s/_index" % rel.rstrip("/").split("/", 1)[-1]) not in grand:
            issues.append(
                "grand index %s has no row for dimension %s/"
                % (profile.grand_index, dim)
            )
    return issues


def lint_report(vault):
    """Human-readable report, one line per issue (matches lint.ts output)."""
    return [i["detail"] for i in lint_vault(vault)]
